src/Tasks/Microsoft.NET.Build.Tasks/ReferenceInfo.cs

```diff
@@ -54,17 +54,36 @@ public static IEnumerable<ReferenceInfo> CreateReferenceInfos(IEnumerable<ITaskI
         public static IEnumerable<ReferenceInfo> CreateDirectReferenceInfos(
             IEnumerable<ITaskItem> referencePaths,
             IEnumerable<ITaskItem> referenceSatellitePaths,
-            bool projectContextHasProjectReferences,
+            LockFileLookup lockFileLookup,
             Func<ITaskItem, bool> isRuntimeAssembly)
         {
-
-            bool filterOutProjectReferenceIfInProjectContextAlready(ITaskItem referencePath)
+            bool lockFileContainsProject(ITaskItem referencePath)
             {
-                return (projectContextHasProjectReferences ? !IsProjectReference(referencePath) : true);
+                if (lockFileLookup == null)
+                {
+                    return false;
+                }
+
+                if (!IsProjectReference(referencePath))
+                {
+                    return false;
+                }
+
+                string projectName = referencePath.GetMetadata(MetadataKeys.MSBuildSourceProjectFile);
+                if (string.IsNullOrEmpty(projectName))
+                {
+                    projectName = Path.GetFileNameWithoutExtension(referencePath.ItemSpec);
+                    if (string.IsNullOrEmpty(projectName))
+                    {
+                        return true;
+                    }
```

Comment:
Under what conditions would this be null or empty, and should we be returning true here?